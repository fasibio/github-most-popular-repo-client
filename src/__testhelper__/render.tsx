import { ReactElement } from 'react';
import {
  render as realRender,
  fireEvent,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';

import { Grommet } from 'grommet';

export const render = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => {
  return realRender(<Grommet>{ui}</Grommet>, options);
};
