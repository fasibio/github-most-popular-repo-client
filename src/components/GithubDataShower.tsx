import { Box, Heading, Spinner, Text, Tabs, Tab, ResponsiveContext } from 'grommet';
import React, { useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useGithubDataLoader } from '../store/github-data';
import { GithubRepoList } from './GithubRepoList';

export const GithubDataShower = () => {
  const { data, load, toggleOnFavoriteList, favoriteIds, favorite } =
    useGithubDataLoader();

  const { t } = useTranslation();
  const size = React.useContext(ResponsiveContext);
  useEffect(() => {
    load();
  }, []);

  if (!data) {
    return (
      <Box direction="row" gap="small" align="center" justify="center" margin={'medium'}>
        <Spinner color={{ dark: 'red', light: 'green' }} />
        <Text size="xxlarge">{t('githubdatashower.loading')}</Text>;
      </Box>
    );
  }
  return (
    <Box
      margin={{ vertical: 'small', horizontal: size !== 'small' ? 'xlarge' : undefined }}
      align={'center'}
    >
      <Heading>{t('githubdatashower.title')}</Heading>
      <Tabs>
        <Tab title={t('githubdatashower.tabs.0.title')}>
          <GithubRepoList
            data={data.items}
            favoriteIds={favoriteIds}
            toggleOnFavoriteList={toggleOnFavoriteList}
          />
        </Tab>
        <Tab title={`${t('githubdatashower.tabs.1.title')} (${favoriteIds.length})`}>
          <GithubRepoList
            data={favorite()}
            favoriteIds={favoriteIds}
            toggleOnFavoriteList={toggleOnFavoriteList}
          />
        </Tab>
      </Tabs>
    </Box>
  );
};
