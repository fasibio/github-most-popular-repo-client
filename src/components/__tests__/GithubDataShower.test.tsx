import {
  GithubRepoState,
  MostStarredRepoResult,
  Repo,
  useGithubDataLoader,
} from '../../store/github-data';
import { GithubDataShower } from '../GithubDataShower';
import type { PartialDeep } from 'type-fest';
import { render } from '../../__testhelper__/render';

jest.mock('../../store/github-data', () => ({
  useGithubDataLoader: jest.fn(),
}));
jest.mock('../GithubRepo.tsx', () => {
  return {
    GithubRepo: (props: any) => {
      return <h1 {...props}>GithubRepo</h1>;
    },
  };
});
describe('tests githubdatashower', () => {
  const renderTests: {
    name: string;
    input: {
      data: PartialDeep<MostStarredRepoResult> | undefined;
      load: jest.Mock;
      toggleOnFavoriteList: jest.Mock;
      favoriteIds: number[];
      favorite: jest.Mock;
    };
    want: {
      loadCalled: boolean;
      toggleOnFavoriteListCalled: boolean;
      favoriteCalled: boolean;
    };
  }[] = [
    {
      name: 'data not downloaded yet show loading screen',
      input: {
        data: undefined,
        favorite: jest.fn(),
        favoriteIds: [],
        load: jest.fn(),
        toggleOnFavoriteList: jest.fn(),
      },
      want: {
        loadCalled: true,
        favoriteCalled: false,
        toggleOnFavoriteListCalled: false,
      },
    },
    {
      name: 'data are loaded and have no favorite',
      input: {
        data: {
          items: [
            {
              id: 1,
              name: 'p1 ',
            } as Repo,
            {
              id: 2,
              name: 'p2 ',
            } as Repo,
          ],
        },
        favorite: jest.fn(),
        favoriteIds: [],
        load: jest.fn(),
        toggleOnFavoriteList: jest.fn(),
      },
      want: {
        loadCalled: true,
        favoriteCalled: true,
        toggleOnFavoriteListCalled: false,
      },
    },
    {
      name: 'data are loaded and have one favorite',
      input: {
        data: {
          items: [
            {
              id: 1,
              name: 'p1 ',
            } as Repo,
            {
              id: 2,
              name: 'p2 ',
            } as Repo,
          ],
        },
        favorite: jest.fn(),
        favoriteIds: [2],
        load: jest.fn(),
        toggleOnFavoriteList: jest.fn(),
      },
      want: {
        loadCalled: true,
        favoriteCalled: true,
        toggleOnFavoriteListCalled: false,
      },
    },
  ];
  renderTests.forEach((test) => {
    it(`render snapshot => ${test.name}`, () => {
      (useGithubDataLoader as unknown as jest.Mock<GithubRepoState>).mockReturnValue({
        data: test.input.data as MostStarredRepoResult,
        load: test.input.load,
        favorite: test.input.favorite,
        favoriteIds: test.input.favoriteIds,
        toggleOnFavoriteList: test.input.toggleOnFavoriteList,
      });
      const component = render(<GithubDataShower />);
      expect(component.container).toMatchSnapshot();
      (useGithubDataLoader as unknown as jest.Mock<GithubRepoState>).mockReset();
      if (test.want.loadCalled) {
        expect(test.input.load).toBeCalled();
      } else {
        expect(test.input.load).not.toBeCalled();
      }
      if (test.want.favoriteCalled) {
        expect(test.input.favorite).toBeCalled();
      } else {
        expect(test.input.favorite).not.toBeCalled();
      }
    });
  });
});
