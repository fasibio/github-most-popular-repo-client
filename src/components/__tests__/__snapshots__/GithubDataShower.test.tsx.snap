// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests githubdatashower render snapshot => data are loaded and have no favorite 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 cfuxjG"
    >
      <h1
        class="StyledHeading-sc-1rdh4aw-0 bdgJOh"
      >
        githubdatashower.title
      </h1>
      <div
        class="StyledBox-sc-13pk1d4-0 hUSFEE StyledTabs-sc-a4fwxl-2 uzbFk"
        role="tablist"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 jKAhip"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 exoCoy StyledTabs__StyledTabsHeader-sc-a4fwxl-0 cUuna-D"
          >
            <button
              aria-expanded="true"
              aria-selected="true"
              class="StyledButton-sc-323bzc-0 hIPuDF"
              role="tab"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jZDVC StyledTab-sc-1nnwnsb-0 jvjGeN"
              >
                <span
                  class="StyledText-sc-1sadyjn-0 bouNPv"
                >
                  githubdatashower.tabs.0.title
                </span>
              </div>
            </button>
            <button
              aria-expanded="false"
              aria-selected="false"
              class="StyledButton-sc-323bzc-0 hIPuDF"
              role="tab"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 etEeAW StyledTab-sc-1nnwnsb-0 jvjGeN"
              >
                <span
                  class="StyledText-sc-1sadyjn-0 hnjXtp"
                >
                  githubdatashower.tabs.1.title (0)
                </span>
              </div>
            </button>
          </div>
        </div>
        <div
          aria-label="githubdatashower.tabs.0.title Tab Contents"
          class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 fvSfiA"
          role="tabpanel"
        >
          <ul
            class="List__StyledList-sc-130gdqg-0 gzRAfE sc-bcXHqe"
            role="list"
          >
            <li
              class="StyledBox-sc-13pk1d4-0 dtdxsN List__StyledItem-sc-130gdqg-1 cSDWGA"
            >
              <h1
                data-testid="GithubRepo"
                repo="[object Object]"
              >
                GithubRepo
              </h1>
            </li>
            <li
              class="StyledBox-sc-13pk1d4-0 dtdxsN List__StyledItem-sc-130gdqg-1 cSDWGA"
            >
              <h1
                data-testid="GithubRepo"
                repo="[object Object]"
              >
                GithubRepo
              </h1>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`tests githubdatashower render snapshot => data are loaded and have one favorite 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 cfuxjG"
    >
      <h1
        class="StyledHeading-sc-1rdh4aw-0 bdgJOh"
      >
        githubdatashower.title
      </h1>
      <div
        class="StyledBox-sc-13pk1d4-0 hUSFEE StyledTabs-sc-a4fwxl-2 uzbFk"
        role="tablist"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 jKAhip"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 exoCoy StyledTabs__StyledTabsHeader-sc-a4fwxl-0 cUuna-D"
          >
            <button
              aria-expanded="true"
              aria-selected="true"
              class="StyledButton-sc-323bzc-0 hIPuDF"
              role="tab"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jZDVC StyledTab-sc-1nnwnsb-0 jvjGeN"
              >
                <span
                  class="StyledText-sc-1sadyjn-0 bouNPv"
                >
                  githubdatashower.tabs.0.title
                </span>
              </div>
            </button>
            <button
              aria-expanded="false"
              aria-selected="false"
              class="StyledButton-sc-323bzc-0 hIPuDF"
              role="tab"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 etEeAW StyledTab-sc-1nnwnsb-0 jvjGeN"
              >
                <span
                  class="StyledText-sc-1sadyjn-0 hnjXtp"
                >
                  githubdatashower.tabs.1.title (1)
                </span>
              </div>
            </button>
          </div>
        </div>
        <div
          aria-label="githubdatashower.tabs.0.title Tab Contents"
          class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 fvSfiA"
          role="tabpanel"
        >
          <ul
            class="List__StyledList-sc-130gdqg-0 gzRAfE sc-bcXHqe"
            role="list"
          >
            <li
              class="StyledBox-sc-13pk1d4-0 dtdxsN List__StyledItem-sc-130gdqg-1 cSDWGA"
            >
              <h1
                data-testid="GithubRepo"
                repo="[object Object]"
              >
                GithubRepo
              </h1>
            </li>
            <li
              class="StyledBox-sc-13pk1d4-0 dtdxsN List__StyledItem-sc-130gdqg-1 cSDWGA"
            >
              <h1
                data-testid="GithubRepo"
                repo="[object Object]"
              >
                GithubRepo
              </h1>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`tests githubdatashower render snapshot => data not downloaded yet show loading screen 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hdSiWx"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 Tvcri"
      />
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
      />
      <span
        class="StyledText-sc-1sadyjn-0 fFQlho"
      >
        githubdatashower.loading
      </span>
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
      />
      ;
    </div>
  </div>
</div>
`;
